File: projects/native-apps/ios/Shared/MultipartForm.swift
import Foundation

/// A `multipart/form-data` body — the text parts, then one file part — in the
/// order the server's parser expects. Kept apart from any request so the body can
/// be written straight to disk: the share extension hands the background session
/// a file, and never holds a second copy of the content in a memory budget that
/// cannot afford one.
struct MultipartForm {
	struct TextPart: Equatable {
		let name: String
		let value: String
	}

	/// The `filename` attribute is what makes the server treat this part as a file
	/// rather than a text field.
	struct FilePart: Equatable {
		let name: String
		let filename: String
		let bytes: Data
	}

	let boundary: String
	let textParts: [TextPart]
	let filePart: FilePart

	var contentType: String { "multipart/form-data; boundary=\(boundary)" }

	/// Writes the body around `filePart.bytes` rather than into a buffer that
	/// duplicates them — the whole reason the body is file-backed.
	func write(to url: URL) throws {
		FileManager.default.createFile(atPath: url.path, contents: nil)
		let handle = try FileHandle(forWritingTo: url)
		defer { try? handle.close() }
		try handle.write(contentsOf: preamble)
		try handle.write(contentsOf: filePart.bytes)
		try handle.write(contentsOf: epilogue)
	}

	private var preamble: Data {
		var data = Data()
		for part in textParts {
			data.append("--\(boundary)\r\n")
			data.append("Content-Disposition: form-data; name=\"\(part.name)\"\r\n\r\n")
			data.append("\(part.value)\r\n")
		}
		data.append("--\(boundary)\r\n")
		data.append("Content-Disposition: form-data; name=\"\(filePart.name)\"; filename=\"\(filePart.filename)\"\r\n\r\n")
		return data
	}

	private var epilogue: Data {
		var data = Data()
		data.append("\r\n")
		data.append("--\(boundary)--\r\n")
		return data
	}
}

private extension Data {
	mutating func append(_ string: String) {
		append(Data(string.utf8))
	}
}

File: projects/native-apps/android/app/src/main/kotlin/com/readplace/android/core/MultipartForm.kt
package com.readplace.android.core

import java.io.File
import java.io.FileOutputStream

/**
 * A `multipart/form-data` body — the text parts, then one file part — in the
 * order the server's parser expects. Kept apart from any request so the body can
 * be written straight to disk: the share target hands the upload a file, and
 * never holds a second copy of the content in a memory budget that cannot afford
 * one.
 */
class MultipartForm(
	val boundary: String,
	val textParts: List<TextPart>,
	val filePart: FilePart,
) {
	data class TextPart(val name: String, val value: String)

	/**
	 * The `filename` attribute is what makes the server treat this part as a file
	 * rather than a text field.
	 */
	class FilePart(val name: String, val filename: String, val bytes: ByteArray)

	val contentType: String get() = "multipart/form-data; boundary=$boundary"

	fun body(): ByteArray = preamble() + filePart.bytes + epilogue()

	/**
	 * Writes around the content bytes rather than into a buffer that duplicates
	 * them — the whole reason the body is file-backed.
	 */
	fun writeTo(destination: File) {
		val sink = FileOutputStream(destination)
		try {
			sink.write(preamble())
			sink.write(filePart.bytes)
			sink.write(epilogue())
		} finally {
			sink.close()
		}
	}

	private fun preamble(): ByteArray {
		val head = StringBuilder()
		for (part in textParts) {
			head.append("--$boundary\r\n")
			head.append("Content-Disposition: form-data; name=\"${part.name}\"\r\n\r\n")
			head.append("${part.value}\r\n")
		}
		head.append("--$boundary\r\n")
		head.append(
			"Content-Disposition: form-data; name=\"${filePart.name}\"; " +
				"filename=\"${filePart.filename}\"\r\n\r\n",
		)
		return head.toString().toByteArray(Charsets.UTF_8)
	}

	private fun epilogue(): ByteArray = "\r\n--$boundary--\r\n".toByteArray(Charsets.UTF_8)
}
